src/unix/linux.c

```diff
@@ -1341,47 +1341,139 @@ static uint64_t uv__read_uint64(const char* filename) {
-}
+static void uv__get_cgroup1_memory_limits(char* buf, uint64_t* high,
+                                          uint64_t* max) {
+  char filename[4097];
+  char* p;
+  char* p_end;
+
+  /* Find out the path of the memory controller. */
+  p = strchr(buf, ':');
+  while (p != NULL && memcmp(p, ":memory:", 8)) {
+    p = strchr(p, '\n');
+    if (p != NULL)
+      p = strchr(p, ':');
+  }
+
+  if (p != NULL) {
+    p = p + 8;
+    p_end = strchr(p, '\n');
+    if (p_end != NULL)
+      *p_end = '\0';
+
+    /* If the memory controller is mounted, check limits there. */
+    snprintf(filename, sizeof(filename), "/sys/fs/cgroup/memory/%s/", p);
+    if (0 == access(filename, F_OK)) {
+      /* Calls to uv__read_uint64 might return 0 if there was a problem. This is
+       * OK because a return value of 0 signifies that the limit is unknown.
+       */
+      snprintf(filename, sizeof(filename),
+              "/sys/fs/cgroup/memory/%s/memory.soft_limit_in_bytes", p);
+      *high = uv__read_uint64(filename);
 
+      snprintf(filename, sizeof(filename),
+              "/sys/fs/cgroup/memory/%s/memory.limit_in_bytes", p);
+      *max = uv__read_uint64(filename);
 
-uint64_t uv_get_constrained_memory(void) {
+      return;
+    }
+  }
+
+  /* Fall back to the global limits of the memory controller. */
+  *high = uv__read_uint64("/sys/fs/cgroup/memory/memory.soft_limit_in_bytes");
+  *max = uv__read_uint64("/sys/fs/cgroup/memory/memory.limit_in_bytes");
+}
+
+static void uv__get_cgroup2_memory_limits(char* buf, uint64_t* high,
+                                          uint64_t* max) {
   char filename[4097];
-  char buf[1024];
+  char* p;
+
+  p = strchr(buf, '\n');
+  if (p != NULL)
+    *p = '\0';
+
+  p = buf + 4;
+
+  snprintf(filename, sizeof(filename), "/sys/fs/cgroup/%s/memory.max", p);
+  *max = uv__read_uint64(filename);
+
+  snprintf(filename, sizeof(filename), "/sys/fs/cgroup/%s/memory.high", p);
+  *high = uv__read_uint64(filename);
+}
+
+static uint64_t uv__get_cgroup_constrained_memory(char* buf) {
   uint64_t high;
   uint64_t max;
-  char* p;
+
+  /* In the case of cgroupv2, we'll only have a single entry. */
+  if (0 == memcmp(buf, "0::/", 4))
+    uv__get_cgroup2_memory_limits(buf, &high, &max);
+  else
+    uv__get_cgroup1_memory_limits(buf, &high, &max);
+
+  if (high == 0 || max == 0)
+    return 0;
+
+  return high < max ? high : max;
+}
+
+uint64_t uv_get_constrained_memory(void) {
+  char buf[1024];
 
   if (uv__slurp("/proc/self/cgroup", buf, sizeof(buf)))
-    return uv__get_constrained_memory_fallback();
+    return 0;
+
+  return uv__get_cgroup_constrained_memory(buf);
+}
+
+
+static uint64_t uv__get_cgroup1_current_memory(char* buf) {
+  return uv__read_uint64("/sys/fs/cgroup/memory/memory.usage_in_bytes");
+}
 
-  if (memcmp(buf, "0::/", 4))
-    return uv__get_constrained_memory_fallback();
+static uint64_t uv__get_cgroup2_current_memory(char* buf) {
```

Comment:
All these `char* buf` parameters make me kind of twitchy. As a callee, you don't really know how big the buffer is, or if it's nul-terminated.

C89 idioms for addressing that are `char* buf, size_t n` (specify length) or `char (*buf)[1024]` (specify array of exactly 1024 elements).

C99 lets you write `char buf[static 1024]`, meaning an array of _at least_ 1024 elements. Libuv is effectively C11 now so using that idiom would be okay.

Reply:
Didn't know about the C99 syntax, that one works perfectly (i.e. without requiring lots of casts, as seemed required with `char (*buf)[1024]`).